### ISSUE TITLE
Some clarification questions about carry

### ISSUE THREAD

**Issue by robert1003** [sapientinc/HRM#26]
First of all thank you for releasing the source code! This is a very cool paper, I am genuinely surprised that it worked this well. While reading the source code I got a bit confused about the role of carry (probably missing something obvious), thus making this post.

It seems like carry is only initialized once here,
https://github.com/sapientinc/HRM/blob/4047578a02e5deba975c38a1f32392547e66c071/pretrain.py#L220
and is resetted only for halted sequence
https://github.com/sapientinc/HRM/blob/4047578a02e5deba975c38a1f32392547e66c071/models/hrm/hrm_act_v1.py#L242
https://github.com/sapientinc/HRM/blob/4047578a02e5deba975c38a1f32392547e66c071/models/hrm/hrm_act_v1.py#L174-L179

which is at most 16 gradient steps if we use the default config
https://github.com/sapientinc/HRM/blob/4047578a02e5deba975c38a1f32392547e66c071/config/arch/hrm_v1.yaml#L7

does that imply we are passing the hidden state across different examples? i.e.

```python
z = z_init

for e in range(epochs):
    for x, y_true in train_dataloader:
        z, y_hat = hrm(z, x)
        loss = softmax_cross_entropy(y_hat, y_true)
        z = z.detach()

        loss.backward()
        opt.step()
        opt.zero_grad()
```

instead of what is described in the paper (fig. 4)
```python
for x, y_true in train_dataloader:
    z = z_init
    for step in range(N_supervision):
        z, y_hat = hrm(z, x)
        loss = softmax_cross_entropy(y_hat, y_true)
        z = z.detach()

        loss.backward()
        opt.step()
        opt.zero_grad()
```

**Comment by robert1003** [comment]
@narvind2003 not sure if I agree with you on this paragraph

> This implementation detail also directly answers your follow-up question, @robert1003. The effective computational depth for a single, difficult example isn't just N*T, but can be up to M_max * N * T. With the default configs, that's 16 * 2 * 2 = 64 total recurrent steps. This longer effective horizon makes the one-step gradient approximation (which assumes proximity to a fixed point) much more plausible than if the model were only running for 4 steps.

it does run 64 total recurrent step, however they perform gradient update every 4 step
